compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUPackSharedMemoryAlloc.cpp

```diff
@@ -0,0 +1,225 @@
+}
+
+/// Insert barriers and wait operations if there are allocs of a different alias
+/// group before the given alloc.
+static void addBarrier(func::FuncOp funcOp, Operation *alloc,
+                       ArrayRef<Operation *> aliasGroup) {
+  Block *entryBlock = &(*funcOp.getBlocks().begin());
+  bool needBarrier = false;
+  if (alloc->getBlock() != entryBlock) {
+    needBarrier = true;
+  } else {
+    for (Operation &op : entryBlock->getOperations()) {
+      if (&op == alloc) break;
+      if (op.getNumRegions() != 0) {
+        needBarrier = true;
+        break;
+      }
+      if (isa<memref::AllocOp>(&op)) {
+        if (std::find(aliasGroup.begin(), aliasGroup.end(), &op) ==
+            aliasGroup.end()) {
+          needBarrier = true;
+          break;
+        }
+      }
+    }
+  }
+  if (!needBarrier) return;
+  OpBuilder builder(alloc);
+  // TODO: make it a option if needed.
+  bool hasAsyncCopies = true;
+  if (hasAsyncCopies) {
+    Value groupToken = builder.create<nvgpu::DeviceAsyncCreateGroupOp>(
+        funcOp.getLoc(), nvgpu::DeviceAsyncTokenType::get(funcOp.getContext()),
+        SmallVector<Value>());
+    builder.create<nvgpu::DeviceAsyncWaitOp>(funcOp.getLoc(), groupToken,
+                                             builder.getI32IntegerAttr(0));
+  }
+  builder.create<gpu::BarrierOp>(alloc->getLoc());
+}
+
+static int64_t getAllocSize(Operation *op, DataLayout &dataLayout) {
+  auto allocOp = cast<memref::AllocOp>(op);
+  int64_t numElements = allocOp.getType().getNumElements();
+  return (dataLayout.getTypeSizeInBits(allocOp.getType().getElementType()) *
+          numElements) /
+         8;
+}
+
+namespace {
+
+struct LLVMGPUPackSharedMemoryAllocPass
+    : public LLVMGPUPackSharedMemoryAllocBase<
+          LLVMGPUPackSharedMemoryAllocPass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<nvgpu::NVGPUDialect>();
+  }
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
+    DominanceInfo dominators(funcOp);
+    SmallVector<Operation *> allocs;
+    funcOp.walk([&](memref::AllocOp alloc) {
+      if (hasSharedMemoryAddressSpace(alloc.getType())) {
+        allocs.push_back(alloc);
+      }
+    });
+    // First sink the alloc as low as possible in the CFG.
+    sinkOpsInCFG(allocs, dominators);
+    SmallVector<AliasGroup> aliasGroups;
+    analyzeSharedMemoryAlloc(funcOp, allocs, aliasGroups);
+    // If there is 1 or less alias group there is nothing to do.
+    if (aliasGroups.size() <= 1) return;
+
+    // Pack all the allocations into one i8 alloc.
+    // We may need to add extra barriers to make sure we are done writting or
+    // reading from the previous alias group before starting a new one.
+    for (size_t i = 0; i < aliasGroups.size(); i++) {
+      for (Operation *alloc : aliasGroups[i]) {
+        addBarrier(funcOp, alloc, aliasGroups[i]);
+      }
+    }
+
+    OpBuilder builder(funcOp.getContext());
```

Comment:
Lines 183 - end of function can be a separable utility function that doesnt need to be GPU specific. The only thing here is the address space.... We can enforce in the `analyse*` that all address spaces for the coalesced memory match. Then for each allocation we can use the address space of any one of the coalasced allocations.